src/legacy/server/index_patterns/routes.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { first } from 'rxjs/operators';
+import { schema } from '@kbn/config-schema';
+import { InternalCoreSetup, KibanaRequest, RequestHandlerContext } from '../../../core/server';
+import { IndexPatternsService } from './service';
+
+export function registerRoutes(core: InternalCoreSetup) {
+  const getIndexPatternsService = async (request: KibanaRequest): Promise<IndexPatternsService> => {
+    const client = await core.elasticsearch.dataClient$.pipe(first()).toPromise();
+    const callCluster = (endpoint: any, params: any, options?: any) =>
+      client.asScoped(request).callAsCurrentUser(endpoint, params, options);
+    return new Promise(resolve => resolve(new IndexPatternsService(callCluster)));
+  };
+
+  const router = core.http.createRouter('/api/index_patterns');
+  router.get(
+    {
+      path: '/_fields_for_wildcard',
+      validate: {
+        query: schema.object({
+          pattern: schema.string(),
+          meta_fields: schema.string({ defaultValue: '[]' }),
+        }),
+      },
+    },
+    async (context: RequestHandlerContext, request: any, response: any) => {
+      const indexPatterns = await getIndexPatternsService(request);
+      const { pattern, meta_fields: metaFields } = request.query;
+
+      let parsedFields = [];
+      try {
+        parsedFields = JSON.parse(metaFields);
+      } catch (error) {
+        return response.badRequest();
+      }
+
+      const fields = await indexPatterns.getFieldsForWildcard({
+        pattern,
+        metaFields: parsedFields,
+      });
+
+      return response.ok({
+        body: { fields },
+        headers: {
+          'content-type': 'application/json',
+        },
+      });
+    }
+  );
+
+  router.get(
+    {
+      path: '/_fields_for_time_pattern',
+      validate: {
+        query: schema.object({
+          pattern: schema.string(),
+          interval: schema.maybe(schema.string()),
+          look_back: schema.number({ min: 1 }),
+          meta_fields: schema.string({ defaultValue: '[]' }),
```

Comment:
does `schema.arrayOf(schema.string())` not work? Surprised to see the change from array of strings to string + the parsing below.  I'm wondering if this could break BWC (haven't checked yet).

Reply:
Currently, new http router doesn't parse JSON-formatted query param and send it as raw string to the handler. So, when we change it to `schema.arrayOf(schema.string(), { defaultValue: [] })`, api_integration tests will fail because the server recognizes JSON string as a single `string` value and responses with 400(Bad request). 

That's why there are `JSON.parse` in both routes. I thought this was an intended behavior. 

If it wasn't, then we need to fix handler to check if string is in JSON format and parse it in the server.

Reply:
Besides, server doesn't catch 404 error even if we create a Boom error object and throw it. So, I had to handle it manually with `try/catch` statement in `_fields_for_time_pattern`. 

As for now, `_fields_for_wildcard` passes all tests because there is no 404 test for it. Because of that, I've asked about it above. It seems to me that we might need to check 404 for `_fields_for_wildcard` and add related test.

Reply:
interesting! So the original router required arrays to be `JSON.stringified` before being sent over, while the new router will do the conversion on both ends. So old code had type`Joi.array().items(Joi.string()).default([]),` but clients still sent a string. So seems like all new router code will have to use `schema.string` and parse manually to avoid breaking BWC...

Does this sound right @joshdover ?

Old way, using `Joi.array().items(Joi.string()).default([]),`:
`string` -> `Array<String>` 
but new router with `schema.arrayOf(schema.string(), { defaultValue: [] })` will do
`Array<string>` -> `Array<string>`
but since we want to preserve BWC, we have to do `schema.string` and manually parse on the client and server.

Reply:
According to the integration test, the new server also accepts if you try to send query param array in `meta_field=_id&meta_field=_source` form. 

So, I changed the schema [like this](https://github.com/elastic/kibana/pull/44993/commits/a544b62cfda942be6515e08ba089855a61908d53#diff-b8f1c4c0cf351dd73576231111548e4aR55):

```ts
          meta_fields: schema.oneOf(
            [
              schema.string({ defaultValue: '[]' }),
              schema.arrayOf(schema.string(), { defaultValue: [] }),
            ],
            { defaultValue: [] }
          ),
```

Reply:
This sounds right. IMO the NP way of doing this is preferred in order to give plugins more flexibility in how they choose to encode query parameters. It's a little more manual work, but I think the less magic here the better.